web/ui/react-app/src/pages/targets/ScrapePoolList.tsx

```diff
@@ -1,61 +1,127 @@
+import { Col, Collapse, Input, InputGroup, InputGroupAddon, InputGroupText, Row } from 'reactstrap';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faSearch } from '@fortawesome/free-solid-svg-icons';
+import { ScrapePoolContent } from './ScrapePoolContent';
+import Filter, { Expanded, FilterData } from './Filter';
 import { useLocalStorage } from '../../hooks/useLocalStorage';
+import styles from './ScrapePoolPanel.module.css';
+import { ToggleMoreLess } from '../../components/ToggleMoreLess';
 
 interface ScrapePoolListProps {
   activeTargets: Target[];
 }
 
-export const ScrapePoolContent: FC<ScrapePoolListProps> = ({ activeTargets }) => {
-  const targetGroups = groupTargets(activeTargets);
+const kvSearch = new KVSearch({
+  shouldSort: true,
+  indexedKeys: ['labels', 'scrapePool', ['labels', /.*/]],
+});
+
+interface PanelProps {
+  scrapePool: string;
+  targetGroup: ScrapePool;
+  expanded: boolean;
+  toggleExpanded: () => void;
+}
+
+export const ScrapePoolPanel: FC<PanelProps> = (props: PanelProps) => {
+  const modifier = props.targetGroup.upCount < props.targetGroup.targets.length ? 'danger' : 'normal';
+  const id = `pool-${props.scrapePool}`;
+  const anchorProps = {
+    href: `#${id}`,
+    id,
+  };
+  return (
+    <div>
+      <ToggleMoreLess event={props.toggleExpanded} showMore={props.expanded}>
+        <a className={styles[modifier]} {...anchorProps}>
+          {`${props.scrapePool} (${props.targetGroup.upCount}/${props.targetGroup.targets.length} up)`}
+        </a>
+      </ToggleMoreLess>
+      <Collapse isOpen={props.expanded}>
+        <ScrapePoolContent targets={props.targetGroup.targets} />
+      </Collapse>
+    </div>
+  );
+};
+
+// ScrapePoolListContent is taking care of every possible filter
+const ScrapePoolListContent: FC<ScrapePoolListProps> = ({ activeTargets }) => {
+  const initialPoolList = groupTargets(activeTargets);
+  const [poolList, setPoolList] = useState<ScrapePools>(initialPoolList);
   const initialFilter: FilterData = {
     showHealthy: true,
     showUnhealthy: true,
   };
   const [filter, setFilter] = useLocalStorage('targets-page-filter', initialFilter);
 
-  const initialExpanded: Expanded = Object.keys(targetGroups).reduce(
+  const initialExpanded: Expanded = Object.keys(initialPoolList).reduce(
     (acc: { [scrapePool: string]: boolean }, scrapePool: string) => ({
       ...acc,
       [scrapePool]: true,
     }),
     {}
   );
   const [expanded, setExpanded] = useLocalStorage('targets-page-expansion-state', initialExpanded);
-
   const { showHealthy, showUnhealthy } = filter;
+
+  const handleSearchChange = (e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
+    if (e.target.value !== '') {
+      const result = kvSearch.filter(e.target.value.trim(), activeTargets);
+      setPoolList(
+        groupTargets(
+          result.map((value) => {
+            return value.original as unknown as Target;
```

Comment:
Do you need two `as...` here?

Reply:
yeah, Typescript is yelling if I'm casting it directly to a Target.